SAK-29036 Bug of special rank: only check part of user eid

Gao-Jun opened the pull request:
Beside changes of code, some SQL statements should be run on existing DB.

teetoppz28 commented:
Looks good to me overall. Thanks for doing this work!

Two things:
1. Please use lower mixed case for variable names, per the Java standard (e.g. "String assignedtodisplay" should be "String assignedToDisplay")
2. I'm not an Oracle expert, so I cannot comment on the Oracle script.

Fix #1 and we should be good to go here...

Gao-Jun commented:
1. All non-standard variable names are only occurs in the code I deleted. In the codes I added or modified are all follow the Java standard.

2. In this patch, I map one record in the old MFR_RANK_T table to multiple records in the new MFR_RANK_INDIVIDUAL_T, and I don't want use any ADVANCED feature (like function or procedure) in SQL. So the SQL may a little complicate. I'd like to explain the ORACLE script as follows:

* **CREATE TABLE MFR_RANK_INDIVIDUAL_T**
New table to keep the records for 'special rank' with users.

* **CREATE TABLE TMP_DIGITS**
Temp table to keep number from 0 to 9.

* **CREATE TABLE TMP_SEQUENCE**
Temp table to keep number from 0 to 99.
In the old MFR_RANK_T, the field of ASSIGNTO is VARCHAR(255), considering it use two characters ('; ') as separator. So, the field only can keep 255/3=81 users maximum. 99 is big enough to handle all existing records.

* **INSERT INTO MFR_RANK_INDIVIDUAL_T**
This is a little complicate.
 * **FROM TMP_SEQUENCE S CROSS JOIN MFR_RANK_T R**
 Make all existing records with 0-99 as out join result
 *  **';' || R.ASSIGNTO||';'**
 Make all existing ASSIGNTO field like ';user1;user2;', start and end with ';'. I'll refer this as 'ASSIGN_TO' in the follow explanation.
 * **INSTR(ASSIGN_TO, ';', 1, S.SEQ)** Get the (S.SEQ)th ';' position of ASSIGN_TO
 * **SUBSTR(ASSIGN_TO, $THE_Nth_POSITION_OF_SEMICOLON, $THE_N+1th_POSITION_OF_SEMICOLON)** Get the Nth user of ASSIGNTO
 * **TRIM (...)** Since the separator is '; ', so the additional blank should be removed.
 * **INNER JOIN SAKAI_USER_ID_MAP M ON TRIM(...)** Get the user id from eid.
 * **1 + LENGTH(R.ASSIGNTO) - LENGTH(REPLACE(R.ASSIGNTO, ';', ''))** Get the number of how many users in this record.
 * **WHERE S.SEQ BETWEEN 1 AND ( 1 + LENGTH(R.ASSIGNTO) - LENGTH(REPLACE(R.ASSIGNTO, ';', '')))** Only extract record as the number of users.
* **ALTER TABLE  MFR_RANK_T ...** Delete the columns which won't be used any more.
* **DROP TABLE ...** Remove temp tables crated before.

teetoppz28 commented:
My apologies, I must have been color blind yesterday... :)

This looks good then!
